### PR TITLE
Add smithy build openapi support

### PR DESCRIPTION
**Pull request by astridej**
As discussed in Discord -

This adds support for supplying a `smithy-build.json` during codegen, either by CLI or by either the SBT or Mill plugins. Initially, the only thing in this file that is supported is the OpenAPI plugin, but in future more features can be parsed and handled in codegen if so desired. Docs have been added that explain this.

This would solve #1478 .

## PR Checklist (not all items are relevant to all PRs)

- [✅ Added unit-tests (for runtime code) _one in the codegen module_
- [👎] Added bootstrapped code + smoke tests (when the rendering logic is modified)
- [✅] Added build-plugins integration tests (when reflection loading is required at codegen-time)
- [👎 ] Added alloy compliance tests (when simpleRestJson protocol behaviour is expanded/updated)
- [👎 ] Updated dynamic module to match generated-code behaviour
- [✅] Added documentation
- [✅] Updated changelog

### CODE REVIEW

File: modules/codegen/src/smithy4s/codegen/internals/CodegenImpl.scala
Comment (astridej) [inline modules/codegen/src/smithy4s/codegen/internals/CodegenImpl.scala:35]: I tried to also handle the existing fields we already had in `SmithyBuild` here (`imports`, `dependencies` and `repositories`) but ran into problems with it, probably because I don't actually understand how the `ModelLoader` works particularly well. 😅 Especially since the PR was big enough already, I figured I'd just sent one implementation that acts on nothing but the OpenApiConfig and one can always iterate further from there.

File: modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala
Comment (astridej) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala]: I had to decide between having a semantically correct model that's easy to use with complex parsing logic, or a model closer to the actual SmithyBuild JSON file in structure that was easy to parse but hard to use. I opted for the first, which means:
* custom `Seq` decoder so the model doesn't have to use `Option[Seq[T]]` to handle missing fields (I believe circe-generic-extras actually has a configuration option for this, but it doesn't seem to be available in all the Scala versions we have here so I went for the homebrewed version)
* modelling the plugins as a sequence, with a hand-written decoder that reads an object of the form
```
{
   "openapi": (openapi config structure goes here),
   "other-plugin": (other plugin structure goes here)
}
```
and looks up the appropriate decoder for each plugin. (Although right now you could model this as a case class with a field `openApi` instead, plugin names can be dynamic so that version wouldn't be future-proof).

Let me know if you think this is too much Circe magic.

File: modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala
Comment (astridej) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala:71]: See comment - as discussed in the Discord, Smithy doesn't seem to think we'd ever need to serialize `OpenApiConfig` objects back to JSON. To be entirely fair to them, we _don't_ - the `writeJson` logic is about creating a `smithy-build.json` file from our SBT/Mill settings, and the openApi configuration isn't one of them. Still, it means that we can't use the same model for read and write.

File: modules/docs/markdown/06-guides/smithy-build-config.md
Comment (astridej) [inline modules/docs/markdown/06-guides/smithy-build-config.md:8]: Questions:
* Does this place make sense for docs, or is there any existing section it could be added to? Nothing seemed to really fit.
* Would it make sense to mention the smithy-build JSON _generation_ capabilities here as well?
* Any other suggestions on how to better document this feature?

File: CHANGELOG.md
Comment (Baccata) [inline CHANGELOG.md]: you'll have to sync with `series/0.18` again, we've released 0.18.16 already

File: modules/codegen-plugin/src/sbt-test/codegen-plugin/smithy-build/smithy-build.json
Comment (Baccata) [inline modules/codegen-plugin/src/sbt-test/codegen-plugin/smithy-build/smithy-build.json:5]: our of curiosity, does this imply that the openapi-translation of only one service can be configured this way ?

File: modules/codegen/src/smithy4s/codegen/internals/CodegenImpl.scala
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/CodegenImpl.scala:35]: that's entirely fine imho : in the context of build-plugins (which constitute the large majority of the smithy4s usage), this information is transmitted over by the build-tool anyway, which should act as the source of truth for this.

File: modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala]: Nah I'm fine with it, thank you 👍

File: modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala:50]: Could you make these implicit values private, please ?

File: modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala
Comment (Baccata) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala]: This certainly works, but I'd rather reconstruct a node by using a Json folder from circe instead of going via the strings.

File: modules/docs/markdown/06-guides/smithy-build-config.md
Comment (Baccata) [reply modules/docs/markdown/06-guides/smithy-build-config.md:8]: > Does this place make sense for docs, or is there any existing section it could be added to? Nothing seemed to really fit.

Yeah I think it's fine. The `Guides` section is  a bit of a misc anyway. 

> Would it make sense to mention the smithy-build JSON generation capabilities here as well? 

I think so ? Even if we end up repeating information

File: modules/codegen-plugin/src/sbt-test/codegen-plugin/smithy-build/smithy-build.json
Comment (astridej) [reply modules/codegen-plugin/src/sbt-test/codegen-plugin/smithy-build/smithy-build.json:5]: The `alloy` conversion logic completely ignores whatever service is set here and will apply [the same OpenAPI config to all services](https://github.com/disneystreaming/alloy/blob/main/modules/openapi/src/alloy/openapi/package.scala#L79-L81). If we wanted to do anything involving service-based filtering, that would have to change first. For now, it's probably worth mentioning this and other Alloy OpenAPI overrides in the docs - I can add a bit about that.

That said, good point that if alloy _did_ check the service name, the current implementation would only allow for one service's OpenAPI to be configured. Checking the Smithy docs, it looks like I overlooked [support for an optional artifact ID after the plugin name](https://smithy.io/2.0/guides/smithy-build-json.html#plugin-id). If I fix that, you'd be able to include separate plugins `openapi::foo` vs `openapi::bar`, but you'd need a change on the alloy side to be able to actually pass multiple of them and have them be respected.

File: modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala
Comment (astridej) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala]: Yeah, I was not really happy about this way of wiring up the smithy `Node` with circe JSON. I'll have a look at the fold logic and see if I can get it to work without going all the way back to the string level.

Comment (astridej) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyBuild.scala]: Done! Number conversion was a little awkward but I _think_ this should minimize any errors introduced by unnecessary rounding.

File: modules/codegen-plugin/src/sbt-test/codegen-plugin/smithy-build/smithy-build.json
Comment (Baccata) [reply modules/codegen-plugin/src/sbt-test/codegen-plugin/smithy-build/smithy-build.json:5]: Alright, that's somewhat disappointing. If it comes to that, I suppose we could codify our own plugin semantics (which could map a unique configuration to services). 

Regardless, this doesn't take away from the fact that this PR is a decent step forward 👍

**Review by Baccata — APPROVED** [review]
Happy with the changes. @kubukoz, if you find the time, would you mind reviewing too ?

**Comment by kubukoz** [comment]
> Happy with the changes. @kubukoz, if you find the time, would you mind reviewing too ?

I'll try closer to the weekend 🥺

**Comment by Baccata** [comment]
@astridej, mind syncing the PR ? I'll merge right after

**Comment by astridej** [comment]
@Baccata Should be up-to-date now!

**Pull request merged**
